File: sdk/block/block_test.go
package block

import (
	"bytes"
	"testing"

	"github.com/bpprotocol/blockparty/sdk/blockpb"
	"github.com/bpprotocol/blockparty/sdk/derive"
	"github.com/bpprotocol/blockparty/sdk/identity"
)

const (
	testWorldPhrase = "bpprotocol.org/v1/test-world"
	testType        = "bpprotocol.org/v1/types/content.post"
	testAudience    = "bpprotocol.org/v1/audience/public-1"
	testTimestamp   = int64(1700000000)
	cosignType      = "bpprotocol.org/v1/cosign.endorse"
)

type fixture struct {
	world    derive.World
	author   identity.Identity
	cosigner identity.Identity
	typeCode derive.Code
	audCode  derive.Code
}

func newFixture() fixture {
	w := derive.OpenWorld(testWorldPhrase)
	return fixture{
		world:    w,
		author:   identity.OpenIdentity(w, "author-pass"),
		cosigner: identity.OpenIdentity(w, "cosigner-pass"),
		typeCode: derive.GetTypeCode(w, testType),
		audCode:  derive.GetAudienceCode(w, testAudience),
	}
}

func (f fixture) signed(data []byte) *blockpb.Block {
	b := New(f.author.Address, f.typeCode, f.audCode, testTimestamp, data)
	Sign(b, f.world, f.author.MLDSA)
	return b
}

func TestSignAndVerify(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))
	if err := Verify(b, f.world, f.author.MLDSA.Public); err != nil {
		t.Fatalf("Verify on a freshly signed block: %v", err)
	}
}

func TestNewSetsContentBindingID(t *testing.T) {
	f := newFixture()
	a := New(f.author.Address, f.typeCode, f.audCode, testTimestamp, []byte("hello"))
	b := New(f.author.Address, f.typeCode, f.audCode, testTimestamp, []byte("hello!"))
	if bytes.Equal(a.Id, b.Id) {
		t.Fatal("block ID did not change with payload")
	}
	if IDHex(a) == "" || len(a.Id) != 32 {
		t.Fatalf("unexpected ID: hex=%q len=%d", IDHex(a), len(a.Id))
	}
}

func TestVerifyMissingSignatures(t *testing.T) {
	f := newFixture()
	b := New(f.author.Address, f.typeCode, f.audCode, testTimestamp, []byte("hi"))
	if err := Verify(b, f.world, f.author.MLDSA.Public); err != ErrMissingSignatures {
		t.Fatalf("Verify unsigned = %v, want ErrMissingSignatures", err)
	}
}

func TestVerifyWrongKeys(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))

	stranger := identity.OpenIdentity(f.world, "stranger")
	if err := Verify(b, f.world, stranger.MLDSA.Public); err != ErrAuthorSignature {
		t.Fatalf("Verify with wrong author key = %v, want ErrAuthorSignature", err)
	}
	otherWorld := derive.OpenWorld("a-different-world")
	if err := Verify(b, otherWorld, f.author.MLDSA.Public); err != ErrWorldSignature {
		t.Fatalf("Verify with wrong world = %v, want ErrWorldSignature", err)
	}
}

func TestTamperingAnyFieldFailsVerification(t *testing.T) {
	f := newFixture()
	cases := []struct {
		name   string
		mutate func(*blockpb.Block)
	}{
		{"version", func(b *blockpb.Block) { b.Version++ }},
		{"id", func(b *blockpb.Block) { b.Id[0] ^= 0xff }},
		{"type_code", func(b *blockpb.Block) { b.TypeCode[0] ^= 0xff }},
		{"audience_code", func(b *blockpb.Block) { b.AudienceCode[0] ^= 0xff }},
		{"timestamp", func(b *blockpb.Block) { b.Timestamp++ }},
		{"data", func(b *blockpb.Block) { b.Data[0] ^= 0xff }},
		{"world_sig", func(b *blockpb.Block) { b.Sigs.World[0] ^= 0xff }},
		{"author_sig", func(b *blockpb.Block) { b.Sigs.Author[0] ^= 0xff }},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			b := f.signed([]byte("hello"))
			c.mutate(b)
			if err := Verify(b, f.world, f.author.MLDSA.Public); err == nil {
				t.Fatalf("tampering %s still verified", c.name)
			}
		})
	}
}

func TestEncodeDecodeRoundTrip(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))
	AddCoSignature(b, cosignType, f.cosigner.MLDSA)

	enc, err := Encode(b)
	if err != nil {
		t.Fatalf("Encode: %v", err)
	}
	b2, err := Decode(enc)
	if err != nil {
		t.Fatalf("Decode: %v", err)
	}
	// Canonical encoding is stable.
	enc2, err := Encode(b2)
	if err != nil {
		t.Fatalf("re-Encode: %v", err)
	}
	if !bytes.Equal(enc, enc2) {
		t.Fatal("re-encoding a decoded block produced different bytes")
	}
	// Signatures survive the round-trip.
	if err := Verify(b2, f.world, f.author.MLDSA.Public); err != nil {
		t.Fatalf("Verify after round-trip: %v", err)
	}
	if cos := CoSignatures(b2); len(cos) != 1 || !VerifyCoSignature(b2, cos[0], f.cosigner.MLDSA.Public) {
		t.Fatal("co-signature did not survive the round-trip")
	}
}

func TestCoSignatureAddVerify(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))
	AddCoSignature(b, cosignType, f.cosigner.MLDSA)

	cos := CoSignatures(b)
	if len(cos) != 1 || cos[0].Type != cosignType {
		t.Fatalf("unexpected co-signatures: %+v", cos)
	}
	if !VerifyCoSignature(b, cos[0], f.cosigner.MLDSA.Public) {
		t.Fatal("valid co-signature failed to verify")
	}
	// Wrong signer key.
	if VerifyCoSignature(b, cos[0], f.author.MLDSA.Public) {
		t.Fatal("co-signature verified under the wrong signer key")
	}
	// The required signatures are unaffected by the presence of a co-signature.
	if err := Verify(b, f.world, f.author.MLDSA.Public); err != nil {
		t.Fatalf("Verify with a co-signature present: %v", err)
	}
}

func TestBadCoSignatureIgnored(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))
	// A garbage co-signature from nobody must not invalidate the block.
	b.Sigs.Extra = append(b.Sigs.Extra, &blockpb.ExtraSignature{
		Type: "evil.example/forge",
		Sig:  []byte("not a real signature"),
	})
	if err := Verify(b, f.world, f.author.MLDSA.Public); err != nil {
		t.Fatalf("a bogus co-signature invalidated the block: %v", err)
	}
	if VerifyCoSignature(b, b.Sigs.Extra[0], f.cosigner.MLDSA.Public) {
		t.Fatal("a bogus co-signature verified")
	}
}

func TestTamperedCoSignatureFailsButBlockValid(t *testing.T) {
	f := newFixture()
	b := f.signed([]byte("hello"))
	AddCoSignature(b, cosignType, f.cosigner.MLDSA)
	b.Sigs.Extra[0].Sig[0] ^= 0xff

	if VerifyCoSignature(b, b.Sigs.Extra[0], f.cosigner.MLDSA.Public) {
		t.Fatal("tampered co-signature verified")
	}
	if err := Verify(b, f.world, f.author.MLDSA.Public); err != nil {
		t.Fatalf("tampered co-signature invalidated the block: %v", err)
	}
}

File: js/packages/sdk/src/block/block.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { bytesToHex, utf8ToBytes } from "../bytes.js";
import { keccak256 } from "../crypto/index.js";
import { openWorld, getTypeCode, getAudienceCode } from "../derive/index.js";
import { openIdentity } from "../identity/index.js";
import {
  newBlock,
  signBlock,
  verifyBlock,
  addCoSignature,
  verifyCoSignature,
  coSignatures,
  encode,
  decode,
  idHex,
  type Block,
} from "./index.js";

const vectors = JSON.parse(
  readFileSync(
    fileURLToPath(new URL("../../../../../conformance/vectors.json", import.meta.url)),
    "utf8",
  ),
) as {
  world: { seed: string };
  block: {
    passphrase: string;
    typeURN: string;
    audienceURN: string;
    timestamp: number;
    data: string;
    idHex: string;
    encodedKeccak: string;
  };
};

const world = openWorld(vectors.world.seed);
const author = openIdentity(world, vectors.block.passphrase);
const typeCode = getTypeCode(world, vectors.block.typeURN);
const audCode = getAudienceCode(world, vectors.block.audienceURN);

function signed(data: Uint8Array): Block {
  const b = newBlock(author.address, typeCode, audCode, vectors.block.timestamp, data);
  signBlock(b, world, author.mldsa);
  return b;
}

describe("cross-impl block fixture (Go ⇄ Node)", () => {
  it("produces a byte-identical signed block to the Go reference", () => {
    const b = signed(utf8ToBytes(vectors.block.data));
    // Same block ID and same canonical encoding fingerprint as Go: since keygen,
    // deterministic ML-DSA signing, and canonical proto encoding all match, the
    // entire signed block is byte-identical across implementations.
    expect(idHex(b)).toBe(vectors.block.idHex);
    expect(bytesToHex(keccak256(encode(b)))).toBe(vectors.block.encodedKeccak);
    expect(verifyBlock(b, world, author.mldsa.publicKey)).toBe(true);
  });
});

describe("sign / verify", () => {
  it("verifies a freshly signed block", () => {
    expect(verifyBlock(signed(utf8ToBytes("hello")), world, author.mldsa.publicKey)).toBe(true);
  });

  it("rejects a missing signature and the wrong keys", () => {
    const unsigned = newBlock(author.address, typeCode, audCode, 1700000000, utf8ToBytes("x"));
    expect(verifyBlock(unsigned, world, author.mldsa.publicKey)).toBe(false);

    const b = signed(utf8ToBytes("hello"));
    const stranger = openIdentity(world, "stranger");
    expect(verifyBlock(b, world, stranger.mldsa.publicKey)).toBe(false);
    const otherWorld = openWorld("a-different-world");
    expect(verifyBlock(b, otherWorld, author.mldsa.publicKey)).toBe(false);
  });

  it("fails verification when any signed field is tampered", () => {
    const mutators: Array<[string, (b: Block) => void]> = [
      ["version", (b) => (b.version += 1)],
      ["id", (b) => (b.id[0] ^= 0xff)],
      ["typeCode", (b) => (b.typeCode[0] ^= 0xff)],
      ["audienceCode", (b) => (b.audienceCode[0] ^= 0xff)],
      ["timestamp", (b) => (b.timestamp += 1n)],
      ["data", (b) => (b.data[0] ^= 0xff)],
      ["world_sig", (b) => (b.sigs!.world[0] ^= 0xff)],
      ["author_sig", (b) => (b.sigs!.author[0] ^= 0xff)],
    ];
    for (const [name, mutate] of mutators) {
      const b = signed(utf8ToBytes("hello"));
      mutate(b);
      expect(verifyBlock(b, world, author.mldsa.publicKey), name).toBe(false);
    }
  });
});

describe("encode / decode", () => {
  it("round-trips canonically and preserves signatures + co-signatures", () => {
    const cosigner = openIdentity(world, "cosigner");
    const b = signed(utf8ToBytes("hello"));
    addCoSignature(b, "bpprotocol.org/v1/cosign.endorse", cosigner.mldsa);

    const enc = encode(b);
    const b2 = decode(enc);
    expect(bytesToHex(encode(b2))).toBe(bytesToHex(enc)); // canonical & stable
    expect(verifyBlock(b2, world, author.mldsa.publicKey)).toBe(true);
    const cos = coSignatures(b2);
    expect(cos.length).toBe(1);
    expect(verifyCoSignature(b2, cos[0]!, cosigner.mldsa.publicKey)).toBe(true);
  });
});

describe("co-signatures", () => {
  it("add/verify works; bad and tampered co-signatures are ignored", () => {
    const cosigner = openIdentity(world, "cosigner");
    const b = signed(utf8ToBytes("hello"));
    addCoSignature(b, "bpprotocol.org/v1/cosign.endorse", cosigner.mldsa);

    const cos = coSignatures(b);
    expect(verifyCoSignature(b, cos[0]!, cosigner.mldsa.publicKey)).toBe(true);
    // Wrong signer key.
    expect(verifyCoSignature(b, cos[0]!, author.mldsa.publicKey)).toBe(false);
    // A co-signature never affects the required signatures.
    expect(verifyBlock(b, world, author.mldsa.publicKey)).toBe(true);

    // Tampered co-signature fails its own check but the block stays valid.
    cos[0]!.sig[0] ^= 0xff;
    expect(verifyCoSignature(b, cos[0]!, cosigner.mldsa.publicKey)).toBe(false);
    expect(verifyBlock(b, world, author.mldsa.publicKey)).toBe(true);
  });
});
